crates/oxc_linter/src/rules/eslint/prefer_const.rs

```diff
@@ -0,0 +1,1062 @@
+    }
+
+    /// Check if a variable should be declared as const
+    fn should_be_const(
+        &self,
+        symbol_id: SymbolId,
+        has_init: bool,
+        is_for_in_of: bool,
+        ctx: &LintContext<'_>,
+    ) -> bool {
+        let symbol_table = ctx.scoping();
+
+        // Get all references to this symbol
+        let references: Vec<_> = symbol_table.get_resolved_references(symbol_id).collect();
+
+        // Count write references (assignments)
+        let write_count = references.iter().filter(|r| r.is_write()).count();
+
+        // If configured to ignore reads before the initial assignment and this variable has an
+        // initializer, then suppress the suggestion when there exists a read that appears before
+        // the declaration. This matches ESLint's `ignoreReadBeforeAssign` behavior for cases like
+        // `class C { static { a; } } let a = 1;` where the read happens prior to the init write.
+        if self.0.ignore_read_before_assign && has_init {
+            let decl_node_id = symbol_table.symbol_declaration(symbol_id);
+            let decl_start = ctx.nodes().get_node(decl_node_id).span().start;
+
+            let has_read_before_decl = references.iter().any(|r| {
+                if !r.is_read() {
+                    return false;
+                }
+                let read_span = ctx.nodes().get_node(r.node_id()).kind().span();
+                read_span.start < decl_start
+            });
+
+            if has_read_before_decl {
+                return false;
+            }
+        }
+
+        // For for-in and for-of loops, the variable gets a new binding on each iteration
+        // If it's never reassigned in the loop body, it should be const
+        if is_for_in_of && write_count == 0 {
+            return true;
+        }
+
+        // If has initializer and no writes, it should be const
+        if has_init && write_count == 0 {
+            return true;
+        }
+
+        // Handle ignoreReadBeforeAssign option
+        if self.0.ignore_read_before_assign && !has_init && write_count == 1 {
+            let write_only_refs: Vec<_> =
+                references.iter().filter(|r| r.is_write() && !r.is_read()).collect();
+
+            if write_only_refs.len() == 1 {
+                let write_ref = write_only_refs[0];
+                let write_node_id = write_ref.node_id();
+
+                // Check if there are any reads before the write
+                let has_read_before_write = references.iter().any(|r| {
+                    if !r.is_read() || r.node_id() == write_node_id {
+                        return false;
+                    }
+                    // Simple span comparison - if read comes before write in source
+                    let read_span = ctx.nodes().get_node(r.node_id()).kind().span();
+                    let write_span = ctx.nodes().get_node(write_node_id).kind().span();
+                    read_span.start < write_span.start
+                });
```

Comment:
how does eslint do this. Relying on spans will be unreliable as functions are hoisted ect.